### ISSUE TITLE
Unit Testing various Kaiaulu Modules

### ISSUE THREAD

**Issue by leilani-reich** [sailuh/kaiaulu#154]
We would like to conduct some unit testing on several functions from social_smell_showcase.md.

Deadline Saturday

Leilani:
Pull Request: https://github.com/sailuh/kaiaulu/pull/162
- [x] parse_mbox()
- [x] parse_jira()

Malia:
Pull Request: https://github.com/sailuh/kaiaulu/pull/163
- [x] git_checkout()
- [x] parse_gitlog()

Nico:
Pull Request: https://github.com/sailuh/kaiaulu/pull/167
- [x] get_date_from_commit_hash()
- [x] filter_by_file_extension()

**Comment by leilani-reich** [comment]
Hi Carlos,

For the unit tests, since I am testing parse_jira_replies(), should I just test parse_jira() as well? Since to get the jira replies, one must call parse_jira_replies(parse_jira(jira_issue_comments_path)).


Also, here are some types of tests ideas I've used so far:
- test that parse_mbox() fails given path to wrong version of Perceval (not 0.12.24) 
- test that I get no errors given I use the correct config file parameters and the correct version of Perceval (0.12.24)
- test that I get an error for passing in an incorrect path to my parse_mbox() or parse_jira_replies()
- test that invalid file type (ex .txt) will fail to be parsed by parse_mbox() or parse_jira_replies()
- test that empty config path(s) passed to parse_mbox() or parse_jira_replies() will cause an error
- test that tables produced by parsers have correct column names (based on the names used in parser.R).

Are there any tests I am missing? In particular, should I test that the values in the table produced by parse_mbox or parse_jira_replies matches the expected output? I was thinking of trying this, but I'm not sure how to get the expected output tables for comparison beforehand.


Finally, is it ok if I define variables at the top of the test file (which I named "test-social-smells.R"), instead of defining the variables in each of the test_that() blocks?

For instance, here I put the variables outside the test_that block:
```r
perceval_path <- "../../../../tools/perceval"
mbox_path <- "../../../../rawdata/mbox/thrift-dev.mbox"
jira_issue_comments_path <- "../../../../rawdata/issue_tracker/thrift_issue_comments.json"

test_that("Wrong version of Perceval fails", {
   expect_error(parse_mbox(perceval_path, mbox_path))
})
```

Thanks,
Leilani

**Comment by carlosparadis** [comment]
Hi,

> For the unit tests, since I am testing parse_jira_replies(), should I just test parse_jira() as well? Since to get the jira replies, one must call parse_jira_replies(parse_jira(jira_issue_comments_path)).

If one function is already called inside another, no need to copy its tests to the other. Was that the case? 

> Are there any tests I am missing? In particular, should I test that the values in the table produced by parse_mbox or parse_jira_replies matches the expected output? I was thinking of trying this, but I'm not sure how to get the expected output tables for comparison beforehand.

I think these tests are fine, but maybe consider them a bit from an analysis standpoint a few more. For example, let's say we want to download issues for a particular day. Does the function succeed in returning an issue with the timestamp on the requested date? Does it return more than what it should? Conversely, if you know a particular day had issues with comments, does the function succeed in retrieving at least one comment?  What about this interaction between the two functions? Is there something we can test when one fails on what happens to the other? Any other parameters the function asks that are not being exercised? 

The empty file is a good test, but also consider how the file looks like if you request a date where only issues with no comments exist, an incorrect JIRA ID, etc. 

I doubt we will ever account for all possible tests, but this is why it is important we add as unit tests the problems encountered, as they are encountered.

> Finally, is it ok if I define variables at the top of the test file (which I named "test-social-smells.R"), instead of defining the variables in each of the test_that() blocks?

Honestly, it has been a while since I last checked the docs for this library. Would you mind checking and seeing what they say about it? If you can't find it anywhere, please let me know too.

**Comment by leilani-reich** [comment]
Hi Carlos,

What I was saying was that the parse_jira function may be called to get the JIRA table and then used as a parameter for parse_jira_replies, since we need JIRA data first before parsing the replies. Actually, I will test parse_jira instead of parse_jira_replies since it's more direct.

My functions parse_mbox and parse_jira don't have date parameters, but I've added a bit more tests to make sure users are getting the correct table output.

Based on https://r-pkgs.org/testing-design.html#sec-testing-design-principles, I think they suggest defining all variables within the testthat blocks, so I will do that.

I will send a pull request for my unit tests today for your review.

Thanks,
Leilani

**Comment by carlosparadis** [comment]
@nicoelee123, any updates on the unit tests for Milestone 1?

**Comment by nicoelee123** [comment]
Hello Carlos,

I was unit testing the function "get_date_from_commit_hash()" which looks like this:
`
get_date_from_commit_hash <- function(git_log,p_commit_hash){
    return(git_log[commit_hash] == p_commit_hash]$author_datetimetz[1])
}
`
I'm testing to see that this function returns "NA" if I pass a table that doesn't have a "commit_hash" column. However, when I do, it still returns the correct date. I'm not sure whether this is expected behavior, and to be honest I'm not sure how it's doing that. I've found that if I change the name of the column to be inside quotation marks, then it acts as expected. For example, if the function was
`
get_date_from_commit_hash <- function(git_log,p_commit_hash){
    return(git_log["commit_hash"] == p_commit_hash]$author_datetimetz[1])
}
`
then it returns "NA" since it cannot find the column "commit_hash". My test looks like this:
`
test_that("NA returned when table doesn't have commit_hash column", {
    minimal_dt = data.table(author_datetimetz, "random_name" = commit_hash, file_pathname)
    output_datetime <- get_date_from_commit_hash(minimal_dt, commit_hash_to_find)
    expect_equal(is.na(output_datetime), TRUE)
})
`
I initialize some variables at the beginning of the test file such as "author_datetimetz", "commit_hash", and "file_pathname". Should I allow this test to fail? Or do I remake the test if this is expected behavior?

**Comment by carlosparadis** [comment]
@nicoelee123 

I've formatted your post a bit so it easier for us to read (take a look on edit mode how it is done with ```r for future reference). 

I am not sure if the way you are creating the new table is correct. Also, on your function where does author_datetimetz comes from? 

Could you give me in plain English a list of the tests you plan to implement?

**Comment by nicoelee123** [comment]
Hello,

For the first concern, the way that I am initializing the tables is by creating lists with names, and then inserting them as columns. For example
```r
letters <- c("a", "b", "c")
numbers <- c(1, 2, 3)

dt = data.table(letters, numbers)
```
This outputs "dt" as

| letters | numbers |
| --- | --- |
| a | 1 |
| b | 2 |
| c | 3 |

If this is incorrect then please let me know. I can also show details on how I did this specifically for my tests by showing the code itself.

Secondly, the tests that I plan on implementing are as follows:

get_date_from_commit_hash()
- Check that the function can correctly return the date associated with a given commit_hash from a table
- Returns NULL when it's given a table with no "author_datetimetz" column
- Returns NA when given a table with no "commit_hash" column
- Returns NA if function cannot find given commit_hash inside given table

filter_by_file_extension()
- Check that function returns correct subset of given table with correct file extensions
- Returns empty table if pathname doesn't exist inside table
- Returns empty if table has no column for path names

**Comment by carlosparadis** [comment]
@leilani-reich @nicoelee123 @malialiu 

Milestone 1 unit tests need some conceptual rework. Here are some summarized notes that may help you 3. Try to write some pseudo-code for me on your respective pull requests (Nico please send me one with your unit tests as soon as you can so you can move to the next milestones too) before making another commit so that we can iterate on it faster and you can keep progressing on Milestone 3. 

###  Sample Git Data on the fly 

I've just pushed a commit to help you with some of the unit tests and clarify what I meant in your reviews as keeping unit tests self-contained: c8c5e02163c9fada80bdb687d54a9f64be5f5bf9 

You want to ensure your unit tests do not require the user to download data in a specific path. They should test what comes on a git clone of Kaiaulu repo, or any assumptions states on the README that the user should follow. 

One difficulty with that is, for example, if you are testing one of the parse functions. For example, `parse_gitlog` since you need a .git file to test them. To that end, I created two functions for you to see how you could go about it. In essence, you want to write a function to SetUp and TearDown the unit tests, by making some mockup lightweight data when the function is called.

This is exactly what the new two functions do. They are fairly simple, and only really perform a set of command line commands via system2 to create a file, add it and make a commit. 

Please use this pair of functions for unit tests associated to git log. Creating a mockup parse_mbox() function is more elaborated, so if your function tests columns on parse_mbox() for example, please map your unit test to test the equivalent on parse_gitlog instead. 

 I did a few quick tests of the function pair on my end, but if you run into any issues let me know. 

A similar idea applies if you are testing something like a project configuration file. You would need to create a sample project configuration file on the fly to then test functions for it. In practice, however, it suffices to test the parameters parse_gitlog() takes directly, since reading from the config files is very simple. 

## Hardcoding filepaths

Remember your test functions should assume a tools.yml exists where it currently is on the git repo in Kaiaulu. If you are testing someone placed Perceval in the correct place, or any other tool, you have to read it from the tools.yml so when the user downloads Kaiaulu, it properly test where they specified it is. If you hardcode the path instead, you are not taking into account the user input, and nothing is being tested.

## Descriptive Names for Tests

Try to be very specific in the description of testthat. When all tests are run, the description should give enough intuition of what you are trying to test and what you expected. Some unit tests only mention what were being tested, but not what was being expected. 

## Precise Assertions 

Try to refine the tests to specific assumptions by asserting equal rather than "return error". For example, if you expect a user to use a version of Perceval, assert equal to the version number the test expects, rather than "something went wrong" when return_error() is used. The assert equals then becomes further documentation of the function, on what behavior it should have given a very specific input. 

Let me know if you have questions here or on your specific pull requests. Thanks!

**Comment by carlosparadis** [comment]
@nicoelee123 

In response to: https://github.com/sailuh/kaiaulu/issues/154#issuecomment-1471337669 

Those all look fine and it seems your unit tests account for very precise assertions too given the input. Your way to create a data.table looks fine for me. I guess at this point, go ahead and make your pull request even with the identified problem, and I will test on my end what the problem is and give you feedback on the PR. I want to try and get you caught up with everyone else on Milestone 3 so you are all on the same page. 

Please also take into consideration my post above. 

Thanks for the quick follow-up!

**Comment by nicoelee123** [comment]
Hello,

As an update, I just submitted the PR for this issue! Please let me know in the PR if there's something I can change or should add/remove.

**Comment by carlosparadis** [comment]
With the last commit in this issue, all unit tests for Milestone 1 are now merged, and thus Milestone 1 is finally done! 🎖️
